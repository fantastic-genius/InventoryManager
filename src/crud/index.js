import AsyncStorage from '@react-native-async-storage/async-storage';
import uuid from 'react-native-uuid';

export const initalizeInventories = async () => {
  try {
    let inventories = await AsyncStorage.getItem('inventories');
    if (!inventories) {
      inventories = [];
      AsyncStorage.setItem('inventories', JSON.stringify(inventories));
    }
  } catch (error) {
    console.log(error);
  }
};

export const addInventory = async inventory => {
  try {
    let inventories = await AsyncStorage.getItem('inventories');
    inventories = JSON.parse(inventories);
    inventories.push({
      uuid: uuid.v4(),
      ...inventory,
    });
    await AsyncStorage.setItem('inventories', JSON.stringify(inventories));
    return inventories;
  } catch (error) {
    console.log(error);
  }
};

export const getInventories = async () => {
  try {
    let inventories = await AsyncStorage.getItem('inventories');
    inventories = JSON.parse(inventories);
    return inventories;
  } catch (error) {
    console.log(error);
  }
};

export const deleteInventory = async inventoryUuid => {
  try {
    let inventories = await AsyncStorage.getItem('inventories');
    inventories = JSON.parse(inventories);
    const newInventories = inventories.filter(
      inventory => inventory.uuid !== inventoryUuid
    );
    await AsyncStorage.setItem('inventories', JSON.stringify(newInventories));
    return newInventories;
  } catch (error) {
    console.log(error);
  }
};

export const updateInventory = async (inventoryUuid, value) => {
  try {
    let inventories = await AsyncStorage.getItem('inventories');
    inventories = JSON.parse(inventories);
    inventories = inventories.map(inventory => {
      if (inventory.uuid === inventoryUuid) {
        return {
          uuid: inventoryUuid,
          ...value,
        };
      }
      return inventory;
    });
    await AsyncStorage.setItem('inventories', JSON.stringify(inventories));
    return inventories;
  } catch (error) {
    console.log(error);
  }
};

export const inventoryNameExist = async (name, inventoryUuid) => {
  try {
    let inventories = await AsyncStorage.getItem('inventories');
    inventories = JSON.parse(inventories);
    let filteredInventories = [];

    if (inventoryUuid) {
      filteredInventories = inventories.filter(
        inventory =>
          inventory.name === name && inventory.uuid !== inventoryUuid
      );
    } else {
      filteredInventories = inventories.filter(
        inventory => inventory.name === name
      );
    }
    
    if (filteredInventories.length) {
      return true;
    }

    return false;
  } catch (error) {
    console.log(error);
  }
};
