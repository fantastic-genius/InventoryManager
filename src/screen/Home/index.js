import React, {useState, useEffect} from 'react';
import {View, StyleSheet, TouchableOpacity, FlatList} from 'react-native';
import Entypo from 'react-native-vector-icons/Entypo';

//constants
import colors from '../../constants/colors';

//components
import Inventory from '../../components/Inventory';

//crud
import {getInventories} from '../../crud';

const Home = ({navigation}) => {
  const [inventories, setInventories] = useState([]);

  const handleGetInvetories = async () => {
    const allInvetories = await getInventories();
    setInventories(allInvetories);
  };

  const renderInventory = ({item}) => (
    <Inventory 
      title={item.name}
      description={item.description}
      price={item.price}
      stock={item.stock}
      key={item.uuid}
      onPress={() => navigation.navigate('EditInventory', {inventory: item})}
      testId={item.uuid}
    />
  );

  useEffect(() => {
    handleGetInvetories();
  }, []);

  return (
    <View style={styles.container}>
      <FlatList
        keyExtractor={item => item.uuid}
        data={inventories}
        renderItem={renderInventory}
        style={styles.listCon}
      />
      <TouchableOpacity
        style={styles.plusCon}
        onPress={() => navigation.navigate('CreateInventory')}
      >
        <Entypo name="plus" size={30} color={colors.WHITE} />
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.BACKGROUND,
  },
  listCon: {
    flexGrow: 1,
    padding: 16,
  },
  plusCon: {
    width: 60,
    height: 60,
    position: 'absolute',
    bottom: 40,
    right: 20,
    backgroundColor: colors.PRIMARY,
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 30,
  },
});

export default Home;
