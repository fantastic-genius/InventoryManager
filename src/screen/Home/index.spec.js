import React from 'react';
import renderer from 'react-test-renderer';
import {render, fireEvent, act} from '@testing-library/react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import Home from './index';
import * as crud from '../../crud';

describe('Home', () => {
  const inventories = [
    {
      uuid: 'hshdjdj-ndnjj-jjfj-fjjjjf',
      name: 'bag',
      desciption: 'bag for school students',
      price: 4000,
      stock: 23,
    },
  ];
  beforeAll(async () => {
    await AsyncStorage.setItem('inventories', JSON.stringify(inventories));
  });

  afterAll(async () => {
    await AsyncStorage.setItem('inventories', JSON.stringify([]));
  });

  it('renders screen as expected', async () => {
    let component;
    renderer.act(async () => {
      component = renderer.create(<Home />);
    });
    component.toJSON();
    expect(component).toMatchSnapshot();
  });

  // it('navigates to edit screen when an inventory is clicked', async () => {
  //   const navigate = jest.fn();
  //   jest.mock('../../crud');
  //   crud.getInventories = jest.fn(() => inventories);
  //   const component = render(<Home />);
  //   const {getByTestId} = component;
  //   expect(crud.getInventories).toHaveBeenCalled();
  //   fireEvent.press(getByTestId(inventories[0].uuid));
  //   expect(navigate).toHaveBeenCalledWith('EditInventory', {inventory: inventories[0]});
  // });
});
