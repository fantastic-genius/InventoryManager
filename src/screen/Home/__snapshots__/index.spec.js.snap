// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders screen as expected 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#E5E5E5",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flexGrow": 1,
        "padding": 16,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#7E42F5",
        "borderRadius": 30,
        "bottom": 40,
        "height": 60,
        "justifyContent": "center",
        "opacity": 1,
        "position": "absolute",
        "right": 20,
        "width": 60,
      }
    }
  >
    <Entypo
      color="#FFFFFF"
      name="plus"
      size={30}
    />
  </View>
</View>
`;
